flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/trait/ModifyKindSet.java

```diff
@@ -0,0 +1,192 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.trait;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * The set of modify operations contained in a changelog.
+ *
+ * @see ModifyKind
+ */
+public class ModifyKindSet {
+
+	/**
+	 * Insert-only modify kind set.
+	 */
+	public static final ModifyKindSet INSERT_ONLY = ModifyKindSet.newBuilder()
+		.addContainedKind(ModifyKind.INSERT)
+		.build();
+
+	/**
+	 * A modify kind set contains all change operations.
+	 */
+	public static final ModifyKindSet ALL_CHANGES = ModifyKindSet.newBuilder()
+		.addContainedKind(ModifyKind.INSERT)
+		.addContainedKind(ModifyKind.UPDATE)
+		.addContainedKind(ModifyKind.DELETE)
+		.build();
+
+	private final Set<ModifyKind> kinds;
+
+	private ModifyKindSet(Set<ModifyKind> kinds) {
+		this.kinds = Collections.unmodifiableSet(kinds);
+	}
+
+	public Set<ModifyKind> getContainedKinds() {
+		return kinds;
+	}
+
+	public boolean contains(ModifyKind kind) {
+		return kinds.contains(kind);
+	}
+
+	public boolean containsOnly(ModifyKind kind) {
+		return kinds.size() == 1 && kinds.contains(kind);
+	}
+
+	public boolean isInsertOnly() {
+		return containsOnly(ModifyKind.INSERT);
+	}
+
+	public int size() {
+		return kinds.size();
+	}
+
+	public boolean isEmpty() {
+		return kinds.isEmpty();
+	}
+
+	/**
+	 * Returns a new set of ModifyKind which is the difference between two sets.
+	 * It is also equal to {@code this.kinds - that.kinds}. For example:
+	 * [I,U,D] diff [I] = [U,D]
+	 * [I,U] diff [U,D] = [I]
+	 * [I,U,D] diff [I,U,D] = []
+	 */
+	public ModifyKindSet diff(ModifyKindSet other) {
+		Set<ModifyKind> result = EnumSet.noneOf(ModifyKind.class);
+		result.addAll(this.kinds);
+		result.removeAll(other.kinds);
+		return new ModifyKindSet(result);
+	}
+
+	/**
+	 * Returns a new ModifyKindSet with all kinds set in both this set and in another set.
+	 */
+	public ModifyKindSet intersect(ModifyKindSet other) {
+		Builder builder = new Builder();
+		for (ModifyKind kind : other.getContainedKinds()) {
+			if (this.contains(kind)) {
+				builder.addContainedKind(kind);
+			}
+		}
+		return builder.build();
+	}
+
+	/**
+	 * Returns a new ModifyKindSet with the union of the other ModifyKindSet.
+	 */
+	public ModifyKindSet union(ModifyKindSet other) {
+		return union(this, other);
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		ModifyKindSet that = (ModifyKindSet) o;
+		return Objects.equals(kinds, that.kinds);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(kinds);
+	}
+
+	@Override
+	public String toString() {
+		if (kinds.isEmpty()) {
```

Comment:
I think we should protect this class to let it be always non-empty.
Does an empty `ModifyKindSet` make any sense?

Reply:
This is on purpose. Sink will produce emtpy `ModifyKindSet`, because it output nothing.